core/src/main/java/io/opencensus/tags/TaggingState.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2017, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.tags;
+
+/** State of the {@link TagsComponent}. */
+public enum TaggingState {
```

Comment:
I think it is clear if you add a third state NO_OP and describe that as a state which cannot be modified and cannot get into that state.

What do you think?

Reply:
You're right, users would probably want to know whether calling `setState(TaggingState.ENABLED)` will have an effect.

Reply:
You're right, users would probably want to know whether calling `setState(TaggingState.ENABLED)` would have an effect.  I'm not sure how we should represent it though.  The third state would be a valid output for getState, but not a valid input for setState.

Reply:
I think we should just ignore on noop maybe linking to where noop is controlled. noop state shouldn't cause damage even to callers of this.

Reply:
> I think we should just ignore on noop maybe linking to where noop is controlled. noop state shouldn't cause damage even to callers of this.

That makes sense, especially since we will also ignore setting ENABLED/DISABLED when the state is NO_OP.

We could also have a separate method for checking whether the tagging implementation is no-op.  Then TaggingState could just represent the different levels of tagging functionality.

Reply:
I added a method `boolean isTaggingAvailable()` to Tags and TagsComponent.  I tried both options, but I thought that it was easier to document the new method than a new enum constant.  What do you think of this design?

Reply:
I personally think that it is wrong to have conceptually one state split into actually 2 APIs "is available" and "set/get state".

I think it is easier to have 3 states as I explained in a previous comment and explain possible transitions that can happen (as a state machine).

Why do you think it was hard to explain? I would explain this way:
* NO_IMPLEMENTATION. This is a permanent state that can be change only by recompiling the library and include the implementation.
* ENABLE. This is the default state of the library and transition from ENABLE to DISABLE can be made via setState. (also include what enable means, propagate tags, etc.)
* DISABLE. same as ENABLE

What do you think? Also we may make SetState throw a checked exception that force people to catch it and take the correct action (it will not be used that often, I don't expect to have more than 1 call to the SetState per binary/library).

Reply:
Ok. Let's do this way. Can you add the same changes to tracing library later?

Reply:
Sure, I opened an issue for the change: #673